### ISSUE TITLE
Add Self-Supervised Contrastive Pretraining for Cold-Start Wallet Embeddings

### ISSUE THREAD

**Issue by Inkman007** [Ledger-Lenz/Ledgerlens-core#314]
## Summary

New wallets produce structurally weak features across the entire `WALLET_GRAPH_FEATURE_NAMES` group in `detection/feature_engineering.py`: `network_centrality` returns `0.0` for any wallet with an empty or single-trade history, `account_age_days` returns `0.0` when creation metadata is unknown, and ring-membership features are trivially zero until the wallet has enough trades to appear in an SCC or GNN-detected ring. This is the classic cold-start problem, and today LedgerLens has no mitigation for it beyond letting these features sit at zero until enough activity accumulates. This issue adds self-supervised contrastive pretraining of wallet embeddings on the unlabelled trade graph, so a new wallet's embedding reflects its position in the graph topology from its very first trade rather than only after enough labelled/behavioural signal has built up.

## Background & Context

`detection/feature_engineering.py`'s `network_centrality(trades, account)` (line ~440) is explicit about the degenerate case: it computes unique counterparties divided by `(total accounts - 1)`, and returns `0.0` immediately if the account has no trades in the current window — a brand-new wallet's first trade produces `network_centrality=0.0` regardless of who it traded with, because centrality is a purely local ratio computed within one window with no memory of prior structure. `account_age_days(account, as_of, account_metadata)` similarly returns `0.0` whenever `account_metadata.get(account, {}).get("created_at")` is missing, which is the default state for any wallet `ingestion/account_loader.py` hasn't yet backfilled funding-source metadata for. `graph_ring_features` returns an all-zero dict (`wash_ring_membership=0.0`, `wash_ring_size=0.0`, `cycle_volume_ratio=0.0`, `timing_tightness_score=0.0`) for any wallet not yet present in `ring_membership` — which requires enough trade history for SCC (`detection/graph_engine.py`) or the GNN (`detection/gnn_ring_detector.py`) to have detected a ring in the first place.

The practical consequence: a wash-trading ring's newest member — the wallet an operator just spun up specifically to route the next leg of a laundering cycle — is the *least* likely to be flagged by the current feature set, precisely because it is new. This is backwards from a detection-priority standpoint; new wallets participating in coordinated trading patterns are exactly what an adversarial operator would prefer regulators/detectors miss.

Self-supervised contrastive pretraining (SimCLR-style: two augmented views of the same wallet's local subgraph should embed close together; augmented views of different wallets' subgraphs should embed far apart) learns general-purpose wallet representations from the *unlabelled* trade graph — no wash/clean labels needed, so it can be trained on the full production trade graph, not just the synthetic labelled data `detection/model_training.py` currently trains the supervised ensemble on. A wallet's very first trade already places it in a local subgraph (its counterparty, that counterparty's other neighbours, etc.) that a pretrained encoder can embed meaningfully, even though `network_centrality` computed over that wallet's own trade count alone is uninformative. This directly complements — not replaces — the existing `GraphSAGEEncoder` (`detection/gnn_ring_detector.py`): the contrastive-pretrained weights become the *initialization* for that encoder, rather than random initialization or training from scratch only on labelled ring/non-ring examples.

## Objectives

- [ ] Implement `GraphAugmenter` producing two augmented views of a wallet's k-hop subgraph via edge dropout and subgraph sampling (SimCLR-style augmentation pair), reusing `detection/graph_engine.py`'s `TradeGraph` for neighbourhood extraction
- [ ] Implement `ContrastiveWalletEncoder` (same `GraphSAGEEncoder` architecture as `detection/gnn_ring_detector.py`, so pretrained weights load directly into the existing encoder) trained with an NT-Xent (normalized temperature-scaled cross-entropy) contrastive loss over augmented view pairs
- [ ] Implement `scripts/pretrain_wallet_embeddings.py` — trains on the full unlabelled production trade graph (not the synthetic labelled dataset), saves a checkpoint `models/wallet_contrastive_encoder.pt` with SHA-256 checksum matching the existing model-integrity pattern
- [ ] Wire `scripts/train_gnn.py` to optionally initialize `GraphSAGEEncoder` from the contrastive-pretrained checkpoint (`--init-from-pretrained models/wallet_contrastive_encoder.pt`) before supervised fine-tuning on labelled ring/non-ring examples
- [ ] Add a `cold_start_embedding_confidence` feature (0-1, higher = embedding is well-supported by neighbourhood structure vs. relying on pretrained priors alone) to `GNN_FEATURE_NAMES`, computed from the wallet's k-hop neighbourhood size at inference time
- [ ] Add an evaluation harness comparing detection AUC-ROC specifically on **new wallets** (age < `COLD_START_AGE_DAYS`, default 7) between random-init and contrastive-pretrained-init encoders, on the synthetic wash-ring generator's newly-created ring members
- [ ] Document the retraining cadence: contrastive pretraining should refresh periodically (e.g. monthly, or triggered by the existing PSI drift signal in `detection/drift_monitor.py`) since the trade graph's topology evolves

## Technical Requirements

**Augmentation** (`detection/contrastive_pretraining.py`, new module):

```python
class GraphAugmenter:
    def __init__(self, edge_dropout_rate: float = 0.2, subgraph_sample_ratio: float = 0.8): ...

    def augment(self, subgraph: "HeteroData") -> "HeteroData":
        """Returns an augmented view: randomly drops edge_dropout_rate fraction
        of edges, then samples subgraph_sample_ratio fraction of remaining nodes
        via random-walk subgraph sampling anchored at the same root wallet.
        """

def build_two_views(wallet: str, graph: "TradeGraph", k_hops: int = 2) -> tuple["HeteroData", "HeteroData"]:
    """Extracts the wallet's k-hop neighbourhood once, then produces two
    independently augmented views for the contrastive pair.
    """
```

**Contrastive loss (NT-Xent)**:

```python
def nt_xent_loss(z_i: torch.Tensor, z_j: torch.Tensor, temperature: float = 0.5) -> torch.Tensor:
    """z_i, z_j: (batch, embed_dim) — embeddings of the two augmented views
    for the same batch of wallets. Positive pairs are (z_i[k], z_j[k]);
    all other combinations in the batch are negatives.
    """
    batch_size = z_i.shape[0]
    z = F.normalize(torch.cat([z_i, z_j], dim=0), dim=1)
    sim = torch.mm(z, z.T) / temperature
    sim.fill_diagonal_(-1e9)
    positive_idx = torch.cat([
        torch.arange(batch_size, 2 * batch_size),
        torch.arange(0, batch_size),
    ])
    return F.cross_entropy(sim, positive_idx)
```

**Encoder reuse** — `ContrastiveWalletEncoder` must be architecturally identical to `detection/gnn_ring_detector.py`'s `GraphSAGEEncoder` (same `in_channels`/`hidden_channels=128`/`out_channels=64`/`num_layers=3` defaults) specifically so `state_dict()` transfers directly:

```python
encoder = GraphSAGEEncoder(in_channels=NODE_FEATURE_DIM)
pretrained_state = torch.load("models/wallet_contrastive_encoder.pt")
encoder.load_state_dict(pretrained_state, strict=True)
# ... continue with supervised fine-tuning on ring/non-ring labels
```

**Cold-start confidence feature**:

```python
def cold_start_embedding_confidence(wallet: str, graph: "TradeGraph", k_hops: int = 2) -> float:
    """1.0 when the wallet's k-hop neighbourhood has >= COLD_START_MIN_NEIGHBOURS
    distinct wallets; scales down linearly toward 0.0 below that, signalling
    that the embedding leans more on pretrained priors than observed structure.
    """
```

**Configuration** (`.env.example`):

```
CONTRASTIVE_PRETRAIN_ENABLED=false
CONTRASTIVE_EDGE_DROPOUT_RATE=0.2
CONTRASTIVE_SUBGRAPH_SAMPLE_RATIO=0.8
CONTRASTIVE_TEMPERATURE=0.5
CONTRASTIVE_MODEL_PATH=models/wallet_contrastive_encoder.pt
CONTRASTIVE_CHECKSUM_PATH=models/wallet_contrastive_encoder.sha256
COLD_START_AGE_DAYS=7
COLD_START_MIN_NEIGHBOURS=3
```

## Security Considerations

- Contrastive pretraining trains on the **unlabelled production trade graph**, which includes real wallet addresses — the pretraining script and any exported embeddings must follow the same wallet-masking convention in logs as the rest of the codebase; raw per-wallet embedding vectors must never be exposed via any API endpoint, consistent with `docs/gnn_ring_detection.md`'s existing "raw embedding vectors are never exposed" rule
- Model integrity: `wallet_contrastive_encoder.pt` verified against `CONTRASTIVE_CHECKSUM_PATH` before load, same pattern as every other model artifact in this codebase; a checksum mismatch must fall back to random initialization (functionally equivalent to the current behaviour) rather than raising
- An adversary controlling many wallets could attempt to pollute the contrastive pretraining corpus by generating synthetic "normal-looking" subgraph structure around wash-ring wallets to bias the pretrained embedding space. Document this as a known limitation and recommend periodic re-evaluation against the labelled ring dataset (the existing `docs/gnn_ring_detection.md` AUC-ROC ≥ 0.85 bar) after each pretraining refresh — a pretraining run that degrades downstream supervised AUC-ROC should not be promoted, mirroring the existing promote/revert logic in `cli.py retrain-check`

## Testing Requirements

- `tests/test_contrastive_pretraining.py`
- Test: `GraphAugmenter.augment` produces a view with fewer edges than the input (respecting `edge_dropout_rate` within tolerance) and preserves the root wallet node
- Test: `build_two_views` returns two distinct `HeteroData` objects for the same wallet with a fixed random seed producing deterministic, reproducible augmentations
- Test: `nt_xent_loss` is minimized (near zero) when `z_i == z_j` for all pairs and negatives are well-separated; is high when embeddings are random noise
- Test: `ContrastiveWalletEncoder`'s `state_dict()` keys match `GraphSAGEEncoder`'s exactly, so `load_state_dict(strict=True)` succeeds
- Test: `cold_start_embedding_confidence` returns `1.0` for a wallet with a dense neighbourhood and scales toward `0.0` for an isolated new wallet
- Test: corrupt/missing pretrained checkpoint falls back to random-init encoder with a `WARNING`, not an exception
- Evaluation test (marked slow/optional in CI): on the synthetic wash-ring generator with newly-created ring members, contrastive-pretrained-init achieves higher AUC-ROC on wallets younger than `COLD_START_AGE_DAYS` than random-init, by a documented minimum margin

## Documentation Requirements

- New `docs/contrastive_pretraining.md`: motivation (cold-start feature degeneracy, citing the specific `network_centrality`/`account_age_days`/`graph_ring_features` zero-value behaviour), SimCLR-style augmentation strategy, NT-Xent loss explanation, pretraining-then-fine-tuning workflow, retraining cadence recommendation
- Update `docs/gnn_ring_detection.md` to cross-reference contrastive pretraining as an optional initialization step before supervised training
- Update `.env.example` with the new `CONTRASTIVE_*`/`COLD_START_*` variables

## Definition of Done

- [ ] All objectives completed
- [ ] Tests pass (`pytest`)
- [ ] No regressions on existing test suite
- [ ] PR reviewed and approved

## For Contributors

**When applying for this issue, please specify:**
- Your area of specialty (e.g., self-supervised learning, graph representation learning, cold-start recommendation/fraud systems)
- Relevant experience with: contrastive learning (SimCLR, GraphCL, or similar), graph augmentation strategies, PyTorch Geometric
- Your approach or initial thoughts on the implementation
- Estimated time to complete

**Ideal contributor profile:** ML researcher/engineer with hands-on self-supervised graph representation learning experience — GraphCL, GCC, or SimCLR-for-graphs specifically — who understands why cold-start entities are both a modeling challenge and, in an adversarial fraud-detection setting, a deliberate evasion vector.

**Comment by Inkman007** [comment]
Closing as part of a full issue-tracker reset: the project's open issues have been replaced with a new curated set of 125 advanced-level issues covering the codebase end to end. No action needed here; if this item is still relevant, please open a fresh issue referencing it.
